content/blog/purrr-1-1-0-parallel/index.Rmd

```diff
@@ -0,0 +1,223 @@
+
+# Parallel version - just wrap your function with in_parallel()
+mtcars |> map_dbl(in_parallel(\(x) mean(x)))
+#>    mpg    cyl   disp     hp   drat     wt   qsec     vs     am   gear   carb 
+#>  20.09   6.19 230.72 146.69   3.60   3.22  17.85   0.44   0.41   3.69   2.81
+
+# Don't forget to clean up when done
+daemons(0)
+```
+
+The results are identical, but the second version distributes the work across your available CPU cores. For computationally intensive tasks, the performance gains can be dramatic.
+
+## Real-world example: parallel model fitting
+
+Let's look at a more realistic scenario where parallel processing truly shines—fitting multiple models:
+
+```r
+library(purrr)
+library(mirai)
+
+# Set up 4 parallel processes
+daemons(4)
+
+# Define a slow model fitting function
+slow_lm <- function(formula, data) {
+  Sys.sleep(0.1)  # Simulate computational complexity
+  lm(formula, data)
+}
+
+# Fit models to different subsets of data in parallel
+models <- mtcars |>
+  split(mtcars$cyl) |>
+  map(in_parallel(\(df) slow_lm(mpg ~ wt + hp, data = df), slow_lm = slow_lm))
+
+# Extract R-squared values
+models |> 
+  map(summary) |> 
+  map_dbl("r.squared")
+#>         4         6         8 
+#> 0.6807065 0.5889239 0.4970692
+
+daemons(0)
+```
+
+Notice how we pass the `slow_lm` function as an argument to `in_parallel()`—this ensures our custom function is available in the parallel processes.
+
+## Production-ready with mirai
+
+The choice of [mirai](https://mirai.r-lib.org) as the parallel backend wasn't arbitrary.
+[mirai](https://mirai.r-lib.org) is a production-grade async evaluation framework that brings several key advantages:
+
+- **Minimal overhead**: Built on modern networking and concurrency principles
+- **Reliable scheduling**: Leveraging fast inter-process communications locally
+- **Scalable architecture**: From multi-process to distributed computing on HPC clusters
+- **Security**: Offers zero-configuration TLS over TCP for additional assurance
+
+This means your parallel purrr code isn't just fast—it's production-ready.
+
+## Creating self-contained functions
+
+One of the key concepts when using `in_parallel()` is creating self-contained functions. Since your function gets serialized and sent to parallel processes, it needs to be completely independent:
+
+```r
+# ❌ This won't work - external dependencies not declared
+my_data <- c(1, 2, 3)
+map(1:3, in_parallel(\(x) mean(my_data)))
+
+# ✅ This works - dependencies explicitly provided
+my_data <- c(1, 2, 3)
+map(1:3, in_parallel(\(x) mean(my_data), my_data = my_data))
+
+# ✅ Package functions need explicit namespacing
+map(1:3, in_parallel(\(x) vctrs::vec_init(integer(), x)))
+
+# ✅ Or load packages within the function
+map(1:3, in_parallel(\(x) {
+  library(vctrs)
+  vec_init(integer(), x)
+}))
+```
+
+This explicit dependency management might seem verbose, but it ensures your parallel code is reliable and predictable—crucial for production environments.
```

Comment:
Mention that we have made the choice of explicit management of dependencies because automatic management brings the risk of pulling in large objects that could cause puzzling loss of performance?